pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BasePartitionUpsertMetadataManager.java

```diff
@@ -217,6 +226,17 @@ public void replaceSegment(ImmutableSegment segment, @Nullable ThreadSafeMutable
         }
         addOrReplaceSegment((ImmutableSegmentImpl) segment, validDocIds, recordInfoIterator, oldSegment,
             validDocIdsForOldSegment);
+
+        // remove expired primary keys from primary key indexes.
+        if (_upsertTTLStrategy != null) {
+          removeExpiredPrimaryKeys(segment, recordInfoIterator, _upsertTTLStrategy.getTtlInMs());
+          MutableRoaringBitmap validDocIdsSnapshot =
```

Comment:
I don't follow the logic here. Why do we persist the valid doc ids here? The new added segment is almost always having all records valid.

Reply:
Thanks @Jackie-Jiang I think i can fix this logic by the following two changes.
- iterating primary key indexes to remove out-of-dated PKs, instead of iterating the recordInfoIterator 
- make sure the expired PK removal should be done before the new consuming segment might already be created.
Is there anything I missed.

Reply:
> Do we plan to implement the "Remove synchronously" approach described in the design doc? If so, it cannot be done in the replace segment because it is too late, and the new consuming segment might already be created. We will need to add a separate method which is invoked after the segment is getting committed, and before the new segment starts consuming.

@Jackie-Jiang is there some examples that invoked after the segment is getting committed, and before the new segment starts consuming?